File: crates/generate/ess-synth/tests/fixtures/accessor-native.go
package system

import (
	"example.invalid/projection/types/core"
	"testing"
)

func TestNativePresenceAndNestedOptionalIdentity(t *testing.T) {
	var inner *string
	body := core.Body{Status: "ready", Nested: &inner}
	event := core.Arrived{Data: body, Partial: nil, Choice: core.ChoiceGone{Value: "gone"}, Wrapped: core.NewWrapped(core.Wrapper{Body: body})}
	input := Project(event)
	if input.Text != "ready" || input.Partial != nil || input.Choice != nil {
		t.Fatalf("wrong partial mapping: %#v", input)
	}
	if input.Nested == nil || *input.Nested != nil {
		t.Fatal("Some(None) was flattened")
	}
	if input.Deeper == nil || *input.Deeper == nil || **input.Deeper != nil {
		t.Fatal("added Some lost nested identity")
	}
	if input.ThroughWrapper != "ready" || input.Whole.Status != "ready" {
		t.Fatal("newtype or whole value lost")
	}
	event.Partial = &body
	event.Choice = core.ChoiceReady{Value: body}
	event.Data.Nested = nil
	input = Project(event)
	if input.Partial == nil || *input.Partial != "ready" || input.Choice == nil || *input.Choice != "ready" {
		t.Fatal("available path lost")
	}
	if input.Nested != nil || input.Deeper == nil || *input.Deeper != nil {
		t.Fatal("None did not receive exactly one Some")
	}
}

File: crates/generate/ess-synth/tests/fixtures/accessor-native.rs
use projection_types::core::{Arrived, Body, Choice, Wrapped, Wrapper};

#[test]
fn native_presence_and_nested_optional_identity() {
    let body = Body { status: "ready".to_owned(), nested: Some(None) };
    let mut event = Arrived {
        data: body.clone(), partial: None, choice: Choice::Gone("gone".to_owned()),
        wrapped: Wrapped(Wrapper { body: body.clone() }),
    };
    let input = projection_system::project(&event);
    assert_eq!(input.text, "ready");
    assert_eq!(input.partial, None);
    assert_eq!(input.choice, None);
    assert_eq!(input.nested, Some(None));
    assert_eq!(input.deeper, Some(Some(None)));
    assert_eq!(input.through_wrapper, "ready");
    assert_eq!(input.whole, body);
    event.partial = Some(body.clone());
    event.choice = Choice::Ready(body);
    event.data.nested = None;
    let input = projection_system::project(&event);
    assert_eq!(input.partial.as_deref(), Some("ready"));
    assert_eq!(input.choice.as_deref(), Some("ready"));
    assert_eq!(input.nested, None);
    assert_eq!(input.deeper, Some(None));
}
